pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+)
+
+// CreateBlockUpload handles requests for creating block upload sessions.
+func (c *MultitenantCompactor) CreateBlockUpload(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	tenantID, _, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "creating block upload session", "user", tenantID, "block_id", blockID)
```

Comment:
What will create "block upload" do?

I think it should do the following:
- check if block with the ID already exist (look for ANY file under prefix)
- create new marker file in the bucket
- list files with prefix again, and verify there is only new marker file
- we can make marker file random, so that multiple calls to CreateBlockUpload will not overwrite each other's files. But then we rely on listing of files on object store being consistent between calls. I think that's fair.
- this random marker file will be deleted when block upload is finished

WDYT?

Reply:
We were thinking about very similar things! https://github.com/grafana/mimir/pull/1694#discussion_r866029751

Reply:
@pstibrany @andyasp I've now tried to implement what @pstibrany asks for above. WDYT?